### ISSUE TITLE
Feature Request: Image Optimization with Pio

### ISSUE THREAD

**Issue by coastalwhite** [getzola/zola#1429]
Maybe a good feature to have optional automation asset image optimization with [pio](https://github.com/siiptuo/pio). This could really reduce a lot of the size of the resulting website. I was busy creating my own pushing pipeline, so I guess I will just suggest the feature.

I would like to know what others think about this.

Note: I am in no way associated with the pio project, and other projects could also work.

**Comment by Keats** [comment]
It could be a nice part of some automatic pipeline. It should probably do like the `imageproc` module does and save the output in `static` so we don't re-process it on every build though.

**Comment by adworacz** [comment]
This seems like an interesting project. For now I optimize images using `mozjpeg/jpegtran` and `advpng` as part of my website publishing script (basically, `zola build` then find images and pass them to optimizers using `fd`/`find`).

This works pretty well, and has a nice "separation of concerns". Not sure how much benefit is brought by adding `pio` directly to Zola vs encouraging users to add it to their own scripts.

**Comment by Keats** [comment]
If the library is in Rust/not too hard to build and the results are good, it's nice to have this kind of stuff built-in since we already do some image manipulation and I can see many people (including me) using that.

**Comment by BezPowell** [comment]
I’ve just started work on a simple perceptual encoding library that uses image-rs for encoders: [BezPowell/perceptual-image - Codeberg.org](https://codeberg.org/BezPowell/perceptual-image). There's still work to be done, such as adding webp support, but is there still interest in adding something like this to Zola?

I'm envisaging a config setting to toggle this on and off, as well as a max_iterations variable, as it will be much slower than regular encoding. As it uses image-rs, jpegs still aren't progressive, but this seemed like a sensible way of adding the feature in without having to completely rewrite Zola's image processing code or adding in tonnes of extra dependencies.

**Comment by Keats** [comment]
Yeah it would be nice. Although one thing is puts into question: should this be done via the template engine as a fn or some kind of pass after rendering is done, like minifiying. I'm leaning towards the additional pass at the end.

**Comment by BezBIS** [comment]
BezPowell here, just commenting from my work account. I do like the idea of running it as a post-processing step, as that should avoid slowing down `serve`. Doing it that way, however, does create 2 potential problems we would have to work around:

1. Perceptual image compression needs access to the original file to compare against. Users would either need to change all of their calls to `resize_image()` to output lossless files, or zola would need to hijack calls to the function during `serve` to output lossless or 100 quality jpegs. How would we then flag which files need to be post-processed, and which were originally intended to be lossless?
2. We would need to write some kind of logic to ensure files are only re-compressed once, and not on each call to `build`; otherwise, quality could slowly creep down over time.

As an alternative, could we do this via the existing `resize_image()` function in the template engine, but reduce the number of iterations it goes through during `serve`?

It's an interesting problem I hadn't given full thought to. On the one hand, you want to give proper processing time to each image, so the encoder can find the smallest possible file for the given quality. On the other hand, you still want to be able to see changes to the site in a reasonable amount of time while developing.

**Comment by Keats** [comment]
I think processed files should just show up in a folder like processed_files or whatever the name is for `resize_image`. You could set up a glob in the config file and zola would first check the cache to avoid re-processing things, just like resize_image. Then it's up to you if you want to commit it or not.

**Comment by BezPowell** [comment]
Just a quick update to say I have opened 2 draft prs (sorry about the clutter!) with different approaches of implementing the feature. #3113 as a post-processing step, and #3118 as an extension to `process_image`. While I lean slightly towards the latter, I am happy to clean-up and continue with either. Any feedback would be appreciated, as they both have pros and cons from both an end-user, and future maintenance point of view.
